### PR TITLE
Add Book and Category testing

### PR DESCRIPTION
**Pull request by nklimovych**
Screen Shot 2024-05-29 at 13 14 03 PM

### CODE REVIEW

File: src/test/resources/database/insert-data-into-db.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    (3, 'Epic', 'Awesome description...'),
+    (4, 'Drama', 'Awesome description...');
+
+INSERT INTO books_categories (book_id, category_id)
+VALUES
+    (1, 1),
+    (2, 2),
+    (3, 2),
+    (4, 2),
+    (5, 3);
```
Comment (azzilian) [inline src/test/resources/database/insert-data-into-db.sql:22]: Imho add - empty line in the eof

File: src/test/resources/database/delete-data-from-db.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+DELETE FROM books_categories;
+DELETE FROM categories;
+DELETE FROM books;
```
Comment (azzilian) [inline src/test/resources/database/delete-data-from-db.sql:3]: IMHO add - empty line in the eof

File: src/test/java/mate/academy/bookstore/service/impl/CategoryServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        return category;
+    }
+
+    private static CategoryDto getCategoryDto() {
+        CategoryDto categoryDto = new CategoryDto();
+        categoryDto.setId(VALID_CATEGORY_ID);
+        categoryDto.setName(VALID_CATEGORY_NAME);
+        return categoryDto;
+    }
+}
```
Comment (azzilian) [inline src/test/java/mate/academy/bookstore/service/impl/CategoryServiceImplTest.java:161]: empty line in the end of file

File: src/test/resources/application.properties [hunk 4]
```diff
@@ -1,8 +1,6 @@
...
-spring.datasource.driverClassName=org.h2.Driver
-spring.datasource.username=sa
-spring.datasource.password=password
-spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
+spring.datasource.url=jdbc:tc:mysql:8:///book_store
+spring.datasource.username=test
+spring.datasource.password=test
 
 jwt.expiration=300000
 jwt.secret=JustAnotherSuperSecretString1234!
```
Comment (azzilian) [inline src/test/resources/application.properties:6]: empty line in the end of file

**Review by azzilian — APPROVED** [review]
Awesome! Just add empty line in the end of some files

File: src/test/resources/database/insert-data-into-db.sql (hunk 1 above)
Comment (nklimovych) [reply src/test/resources/database/insert-data-into-db.sql:22]: I believe there is an issue with GitHub as it does not display an empty line at the end.

Знімок екрана 2024-05-30 о 00 00 50

File: src/test/resources/database/delete-data-from-db.sql (hunk 2 above)
Comment (nklimovych) [reply src/test/resources/database/delete-data-from-db.sql:3]: Same here

Знімок екрана 2024-05-30 о 00 01 26

File: src/test/java/mate/academy/bookstore/service/impl/CategoryServiceImplTest.java (hunk 3 above)
Comment (nklimovych) [reply src/test/java/mate/academy/bookstore/service/impl/CategoryServiceImplTest.java:161]: Знімок екрана 2024-05-30 о 00 03 20

File: src/test/resources/application.properties (hunk 4 above)
Comment (nklimovych) [reply src/test/resources/application.properties:6]: Знімок екрана 2024-05-30 о 00 02 39

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    @Test
+    @WithMockUser(roles = {USER_ROLE, ADMIN_ROLE})
+    @DisplayName("Get all books (valid request)")
+    void getAllBooks_ValidRequest_ReturnsListOfBooks() throws Exception {
+        when(bookService.findAll(any(Pageable.class))).thenReturn(
+                Collections.singletonList(bookDto));
+
+        mockMvc.perform(get(BASE_URL))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath(TITLE_0_EXPRESSION).value(VALID_BOOK_TITLE));
```
Comment (andrii-hoienko) [inline src/test/java/mate/academy/bookstore/controller/BookControllerTest.java:98]: Better to check full response object (check other cases as well)

File: src/test/java/mate/academy/bookstore/repository/book/BookRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+    @Test
+    @DisplayName("Find a book by isbn - (valid isbn)")
+    @Sql(scripts = {DELETE_DATA_FROM_DB, INSERT_DATA_INTO_DB},
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    void findBookByIsbn_ValidIsbn_ReturnsBookWithMatchingIsbn() {
+        Optional<Book> bookOptional = bookRepository.findBookByIsbn(VALID_BOOK_ISBN_KOBZAR);
+        assertTrue(bookOptional.isPresent());
+        assertEquals(VALID_BOOK_ID_KOBZAR, bookOptional.get().getId());
+        assertEquals(VALID_BOOK_ISBN_KOBZAR, bookOptional.get().getIsbn());
```
Comment (andrii-hoienko) [inline src/test/java/mate/academy/bookstore/repository/book/BookRepositoryTest.java]: CHeck full object ignoring id

Comment (andrii-hoienko) [reply src/test/java/mate/academy/bookstore/repository/book/BookRepositoryTest.java]: CHeck other methods and classes as well

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, left minor comments

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+public class BookControllerTest {
+    private static final String ID = "/{id}";
+    private static final String TITLE_PARAM_NAME = "title";
+    private static final String AUTHOR_PARAM_NAME = "author";
+    private static final String ISBN_PARAM_NAME = "isbn";
+    private static final String PRICE_PARAM_NAME = "price";
+    private static final String BASE_URL = "/books";
+    private static final String SEARCH_URL = BASE_URL + "/search";
+    private static final String ADMIN_ROLE = "ADMIN";
+    private static final String USER_ROLE = "USER";
```
Comment (karma-o) [inline src/test/java/mate/academy/bookstore/controller/BookControllerTest.java]: why don't you use enum values with `name()` method instead?

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    private static final BigDecimal VALID_BOOK_PRICE = BigDecimal.valueOf(14.99);
+    private static final Set<Long> VALID_CATEGORY_ID = Set.of(1L);
+    private static final Long VALID_BOOK_ID = 1L;
+    private static final String TITLE_0_EXPRESSION = "$[0].title";
+    private static final String TITLE_EXPRESSION = "$.title";
+
+    private static MockMvc mockMvc;
+
+    @MockBean
+    private BookService bookService;
```
Comment (karma-o) [inline src/test/java/mate/academy/bookstore/controller/BookControllerTest.java:56]: I think our api interation tests should not use mocked services, but rather a real service and a testcontainer DB. Same for the category controller test

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java (hunk 5 above)
Comment (karma-o) [reply src/test/java/mate/academy/bookstore/controller/BookControllerTest.java:98]: not resolved

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.data.domain.Pageable;
+import org.springframework.http.MediaType;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class BookControllerTest {
```
Comment (karma-o) [inline src/test/java/mate/academy/bookstore/controller/BookControllerTest.java:52]: we can add some invalid cases, like non existent book, same for category test and services

File: src/test/java/mate/academy/bookstore/service/impl/BookServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+        bookDto.setTitle(VALID_BOOK_TITLE);
+        bookDto.setAuthor(VALID_BOOK_AUTHOR);
+        bookDto.setIsbn(VALID_BOOK_ISBN);
+        bookDto.setPrice(VALID_BOOK_PRICE);
+        bookDto.setDescription(VALID_BOOK_DESCRIPTION);
+        bookDto.setCoverImage(VALID_BOOK_COVER_IMAGE);
+        return bookDto;
+    }
+
+    private BookDto getBookDto(CreateBookRequestDto dto) {
```
Comment (karma-o) [inline src/test/java/mate/academy/bookstore/service/impl/BookServiceImplTest.java:266]: it is not just the getter, it also has some parameter, and it essentially maps something to something

**Review by karma-o — APPROVED** [review]
Looks good! Left a few suggestions myself

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java (hunk 5 above)
Comment (nklimovych) [reply src/test/java/mate/academy/bookstore/controller/BookControllerTest.java:98]: I don't understand how I missed that...

File: src/test/java/mate/academy/bookstore/controller/BookControllerTest.java (hunk 7 above)
Comment (nklimovych) [reply src/test/java/mate/academy/bookstore/controller/BookControllerTest.java]: It's hard to say))

Comment (nklimovych) [reply src/test/java/mate/academy/bookstore/controller/BookControllerTest.java]: Maybe I did something wrong, but it won't let me do it this way.
Знімок екрана 2024-06-06 о 21 34 01

**Pull request merged**